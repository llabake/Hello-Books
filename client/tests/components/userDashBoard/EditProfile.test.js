import React from 'react';

import {
  EditProfile,
  mapDispatchToProps,
  mapStateToProps
} from '../../../components/userDashBoard/EditProfile';

import adminData from '../../mocks/userData'

function setup() {
  const props = {
    dispatch: jest.fn(),
    user: adminData.admin,
    errors: {},
    profileToEdit: adminData.userProfileResponse,
    loading: false,
    editProfile: jest.fn(),
    getUserProfile: jest.fn(),
    handleFileChange: jest.fn(),
    handleChange: jest.fn(),
    handleSubmit: jest.fn(),
    handleBlur: jest.fn(),
  };

  const state = {
    image: '',
    errors: {},
    isValid: false,
    saving: false,
    uploadedImage: null,
    redirect: false
  };

  const event = {
    preventDefault: jest.fn(),
    target: {
      name: 'username',
      value: 'username'
    }
  };
  const wrapper = shallow(<EditProfile {...props} {...state} />);

  return {
    wrapper,
    props,
    state,
    event
  };
}


describe('Modify Profile page', () => {
  test('should render with the right amount of elements', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  })

  test('should change state for title', () => {
    const { wrapper, props, event, state } = setup();
    wrapper.find('.signup').simulate('change', event);
    wrapper.instance().handleChange(event)
    expect(wrapper.instance().state.username).toBe('username');
  });

  test('should call handle change', () => {
    const { wrapper, props, state } = setup();
    const event = {
      preventDefault: jest.fn(),
      target: {
        username: 'another username'
      }
    }
    wrapper.setState({
      username: 'another username',
    });
    props.editProfile.mockReset();
    wrapper.instance().handleChange(event)
    expect(wrapper.instance().state.username).toBe('another username');
  });

  test('should call componentWillReceiveProps', () => {
    let spy;
    afterEach(() => {
      spy.mockClear();
    });

    const props = {
      dispatch: jest.fn(),
      user: adminData.admin,
      errors: {},
      profileToEdit: adminData.userProfileResponse,
      editProfile: jest.fn(() => Promise.resolve(1)),
      getUserProfile: jest.fn(() => Promise.resolve(1)),
      updated: false
    }

    const nextProps = {
      dispatch: jest.fn(),
      user: adminData.admin,
      errors: {},
      editProfile: jest.fn(() => Promise.resolve(1)),
      getUserProfile: jest.fn(() => Promise.resolve(1)),
      updated: true
    }
    spy = jest.spyOn(EditProfile.prototype, 'componentWillReceiveProps');
    const wrapper = shallow(<EditProfile {...props} />)
    wrapper.setProps(nextProps);
    expect(EditProfile.prototype.componentWillReceiveProps).toHaveBeenCalledTimes(1)
    expect(spy).toHaveBeenCalled();
  })

  test('should handle form submission when the form inputs are valid', () => {

    const { wrapper, event, props } = setup();
    wrapper.setState({
      username: 'username',
    });
    const dispatchSpy = jest.spyOn(wrapper.instance().props, 'dispatch');
    props.editProfile.mockReset();
    wrapper.instance().handleSubmit(event);
    expect(wrapper.instance().state.username).toBe('username');
    
    // expect(dispatchSpy).toHaveBeenCalledTimes(1);
  });

  test('should handle image upload', () => {
    const { wrapper, props, state } = setup();
    const event = {
      preventDefault: jest.fn(),
      target: {
        files: ['well firts image link',]
      }
    }
    wrapper.setState({
      image: '',
      uploadedImage: event.target.files[0],
    });
    props.editProfile.mockReset();
    wrapper.instance().handleFileChange(event)
    expect(wrapper.instance().state.uploadedImage).toBe('well firts image link');
  });

  test('should dispatch needed actions', () => {
    const dispatch = jest.fn();
    expect(mapDispatchToProps(dispatch)).toHaveProperty('getUserProfile')
    expect(mapDispatchToProps(dispatch)).toHaveProperty('editProfile')
    const { editProfile, getUserProfile } = mapDispatchToProps(dispatch);
    getUserProfile();
    editProfile();
    expect(dispatch).toHaveBeenCalled();
  })

  test('should return an object with props required by component from store', () => {
    const store = {
      errors: {},
    };
    expect(mapStateToProps({ userReducer: {} })).toHaveProperty('profileToEdit');
    expect(mapStateToProps({ userReducer: {} })).toHaveProperty('updated');
  });
})