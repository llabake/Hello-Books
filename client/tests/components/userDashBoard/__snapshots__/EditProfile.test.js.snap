// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modify Profile page should render with the right amount of elements 1`] = `
<div>
  <div
    id="banner"
  >
    <div
      className="container form-style signin-form"
    >
      <div
        className="row"
      >
        <div
          className="row signup-head"
        >
          <div
            className="col s12"
          >
            <h5
              className="center-align"
            >
               Edit Profile 
            </h5>
          </div>
        </div>
        <div
          className="col s12 "
        >
          <div
            className="row "
          >
            <form
              className="col s12 signup"
              onSubmit={[Function]}
            >
              <TextInput
                icon="account_circle"
                id="first_name"
                name="firstName"
                onChange={[Function]}
                placeholder="First Name"
                type="text"
              />
              <TextInput
                icon="account_circle"
                id="last_name"
                name="lastName"
                onChange={[Function]}
                placeholder="Last Name"
                type="text"
              />
              <div
                className="file-field input-field"
                style={
                  Object {
                    "marginTop": 110,
                    "top": "1em",
                  }
                }
              >
                <div
                  className="btn"
                  style={
                    Object {
                      "fontSize": "13px",
                      "marginLeft": "1.8em",
                      "width": "40%",
                    }
                  }
                >
                  <span>
                    Photo
                  </span>
                  <input
                    accept="image/*"
                    onChange={[Function]}
                    type="file"
                  />
                </div>
                <div
                  className="file-path-wrapper"
                >
                  <input
                    className="file-path validate"
                    style={
                      Object {
                        "marginTop": 9,
                        "width": "89%",
                      }
                    }
                    type="text"
                  />
                </div>
              </div>
              <button
                className="waves-effect waves-light btn"
                disabled={true}
                style={
                  Object {
                    "top": "4em",
                  }
                }
                type="submit"
              >
                Save
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
